### ISSUE TITLE
'Capture' quests get cancelled

### ISSUE THREAD

**Issue by SurDno** [Decane/SRP#3]
'Capture' quests that involve a group of stalkers not really capturing point, but instead killing mutants and returning to base (e.g. the quest given by CS squad commanders in Fisherman Farmstead) sometimes get cancelled instead of completed and you do not get reward even if you kill every single mutant.
I have two savefiles reproducing this issue: one with bloodsucker (https://yadi.sk/d/OGiTXMK-3Tphri) and another one with dogs - there is only one remaining, but believe me, I've killed all the others (https://yadi.sk/d/d-GOJf383TphsT).

UPD: As Decane found out further, it does not matter if stalkers capture the point or not, that's a bug related to one-shot kills.

**Comment by Decane** [comment]
Thanks for uploading savegames from which to reproduce the issue. I tested the one with a single dog remaining. Observations:

**(A)** Player does not injure dog, NPC kills dog => **task cancelled**.
**(B)** Player one-shot kills dog => **task cancelled**.
**(C)** Player injures, then kills dog => **task completed**.
**(D)** Player injures, NPC kills dog => **task completed**.

In case (A), the witnessed behavior is as intended, provided that the player also didn't injure any of the other dogs - if the player doesn't "contribute" to the task by injuring or killing any enemies, the task is, and should be, cancelled. However, the behavior witnessed in (B) is wrong, and appears to be hit/death callback related - the `actor_helped` class variable isn't being set to `true` when it should be.

The behavior in case (A) would also be wrong if the player did hit at least one of the other dogs before the last remaining one.

**Comment by SurDno** [comment]
I've killed every single other dog with a shotgun, that made it one-shot. Guess that's B behaviour.
Thanks, now I know that I have to injure them first.

**Comment by Decane** [comment]
I've found the problem and pushed a fix to the master branch. It was caused by a race condition. Details:

Whenever you kill an enemy, hit and death callbacks are supposed to fire—preferably in that order—from the object binder of the killed enemy (xr_motivator.script for stalkers; bind_monster.script for mutants). Additionally, a death callback fires from se_monster.script / se_stalker.script on the server-side killed entity. Ideally, this would fire last.

However, in your savegames, killing the last remaining mutant triggered the server-side `function se_monster:on_death()` from se_monster.script before either of the aforementioned hit or death callbacks from bind_monster.script. `on_death()` in turn called `on_npc_death()` from sim_squad_generic.script, which contained this chunk of code:
```
	local st_npc = db.storage[npc_id]
	if st_npc ~= nil then
		st_npc.squad_obj = nil
	end
```
So by the time the death callback on the server-side entity had finished, `db.storage[npc_id].squad_obj` had been set to `nil` for the killed entity. That's puzzle piece # 1.

Puzzle piece # 2 is as follows: Class variable `actor_helped` is supposed to be set to `true` whenever the player helps out in a 'capture' or 'eliminate' task. ("Helping out" here means inflicting damage to the enemy.) Then, when the task is finished, `actor_helped` is checked to see if the player helped, and if so, the task completes; otherwise, it cancels.

In your case, `actor_helped` was not being set to `true` when firing the lethal shot (and hence the 'capture' task was cancelling instead of completing) because the `actor_help()` call in task_objects.script which is supposed to set `actor_helped` to `true` was never reached. That's because the call to `actor_help()` can only be reached via `sim_combat.hit_callback()` conditional on `db.storage[npc_id].squad_obj` not being `nil`. But that _was_ `nil` per puzzle piece # 1 above.

My solution was to delegate the assignment of `db.storage[npc_id].squad_obj` to `nil` from function `sim_squad_generic:on_npc_death()` to the death callbacks of the object binders, and to precede this assignment with a call to `sim_combat.hit_callback()`. That way, even if the death callback on the server entity gets called first, the `actor_helped` variable will be set correctly before `db.storage[npc_id].squad_obj = nil`.
